Cleanup and modernize SeedingLayerSetsBuilder

makortel opened the pull request:
This PR removes unused and obsolete code related to SeedingLayerSetsBuilder:
* `ctfseeding::SeedingLayer` class altogether (made obsolete by #2378)
   * `TrackingRegion::hits(const EventSetup&, const SeedingLayer *)`
   * The only still-useful piece of `SeedingLayer` was `SeedingLayer::Side` enumeration, which I moved to `DataFormats/TrackerCommon/interface/TrackerDetSide.h` as `TrackerDetSide` enumeration (still "old-style" enum as the numerical values are heavily used)


and does some modernization
* use `unique_ptr` instead of `shared_ptr` (since it's enough)
* add `fillDescriptions` (even if it is incomplete)
   * as a consequence some strip pair layer lists are updated to use the same layer naming convention as in iterative tracking configuration
* improve the relation between `SeedingLayerSetsBuilder` and `SeedingLayerEDProducer`

The last two points are also a preparatory work for the upcoming "automated mitigation for pixel inactive areas in pixel pair iteration".

Tested in 9_4_0_pre3, no changes expected in results except for MTV seeding layer histograms for conversion seeds
* for some layers TIBn/TOBn move to MTIBn/MTOBn
* TECn get split to TECn and MTECn

@VinInn

makortel committed:
Remove SeedingLayerSetsBuilder::layers() as unused and obsolete

makortel committed:
Remove CosmicTrackingRegion::hits(.., ctfseeding::SeedingLayer*) as unused and obsolete

makortel committed:
Remove unnecessary includes of SeedingLayer.h

makortel committed:
Remove all functionality from SeedingLayer as unused and obsolete

makortel committed:
Remove forward declaration

makortel committed:
Move SeedingLayerSetsHits construction to SeedingLayerSetsBuilder

In the near future I plan to use SeedingLayerSetsBuilder::hits()
elsewhere than SeedingLayersEDProducer, and for that I want the
construction of SeedingLayerSetsHits to be solely on the
responsibility of SeedingLayerSetsBuilder.

makortel committed:
Minor modernization

makortel committed:
Deliver also DetLayer vector via pointer for consistency

makortel committed:
Remove SeedingLayer.h as unnecessary, move enum to more general TrackerDetSide

makortel committed:
Remove unnecessary mutable

makortel committed:
Use unique_ptr

makortel committed:
Simplify updating from EventSetup

makortel committed:
Move SeedingLayersEDProducer to fillDescriptions

Still with setAllowAnything(), but better than nothing.

makortel committed:
Unify layer naming with iterative tracking and reduce copy-paste

Towards fillDescriptions in SeedingLayersEDProducer, makes it easier if there are no TID[1-4].

makortel committed:
Call updateEventSetup inside hits() in SeedingLayerSetsBuilder

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21138/1773

cmsbuild commented:
A new Pull Request was created by  @makortel (Matti Kortelainen) for master.

It involves the following packages:

DataFormats/TrackerCommon
RecoMuon/TrackerSeedGenerator
RecoPixelVertexing/PixelLowPtUtilities
RecoPixelVertexing/PixelTrackFitting
RecoPixelVertexing/PixelTriplets
RecoTracker/Configuration
RecoTracker/ConversionSeedGenerators
RecoTracker/SpecialSeedGenerators
RecoTracker/TkSeedingLayers
RecoTracker/TkTrackingRegions
TrackingTools/TransientTrackingRecHit
Validation/RecoTrack


@perrotta, @vazzolini, @kmaeshima, @dmitrijus, @cmsbuild, @slava77, @vanbesien can you please review it and eventually sign? Thanks.
@ghellwig, @trocino, @calderona, @felicepantaleo, @abbiendi, @GiacomoSguazzoni, @rovere, @VinInn, @echapon, @bellan, @HuguesBrun, @jhgoh, @wmtford, @mschrode, @gpetruc, @ebrondol, @rociovilar, @mtosi, @dgulhan, @bachtis, @battibass this is something you requested to watch as well.
@davidlange6, @slava77 you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 2a0997596181c8213618834141d804566253c9f4
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21138/24141/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/67026de8c39bef5dc0a2b1efd113561c78c995b6
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21138/24141/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21138/24141/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21138/24141/summary.html

There are some workflows for which there are errors in the baseline:
 10024.0 step 5
10042.0 step 5
 The results for the comparisons for these workflows could be incomplete 
This means most likely that the IB is having errors in the relvals.The error does NOT come from this pull request

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 203 differences found in the comparisons
 - DQMHistoTests: Total files compared: 26
 - DQMHistoTests: Total histograms compared: 2455601
 - DQMHistoTests: Total failures: 1144
 - DQMHistoTests: Total nulls: 4
 - DQMHistoTests: Total successes: 2454296
 - DQMHistoTests: Total skipped: 157
 - DQMHistoTests: Total Missing objects: 0
 - Checked 107 log files, 10 edm output root files, 26 DQM output files

makortel commented:
> Reco comparison results: 203 differences found in the comparisons

If I interpreted the logs correctly, only the cosmics workflows are affected, I'll take a look tomorrow.

makortel committed:
Fix cosmics

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21138/1787

cmsbuild commented:
Pull request #21138 was updated. @perrotta, @vazzolini, @kmaeshima, @dmitrijus, @cmsbuild, @slava77, @vanbesien can you please check and sign again.

makortel commented:
@cmsbuild, please test

> only the cosmics workflows are affected

Should be fixed now.

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: dba7e34c127f3315b06c92cc18d869b8155bb642
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21138/24148/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/e63f423ffcee397c63651620c3e4f627dc781f83
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21138/24148/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21138/24148/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21138/24148/summary.html

There are some workflows for which there are errors in the baseline:
 10042.0 step 5
 The results for the comparisons for these workflows could be incomplete 
This means most likely that the IB is having errors in the relvals.The error does NOT come from this pull request

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 26
 - DQMHistoTests: Total histograms compared: 2823567
 - DQMHistoTests: Total failures: 30
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2823367
 - DQMHistoTests: Total skipped: 170
 - DQMHistoTests: Total Missing objects: 0
 - Checked 107 log files, 10 edm output root files, 26 DQM output files

dmitrijus commented:
+1

slava77 commented on DataFormats/TrackerCommon/interface/TrackerDetSide.h:
```diff
@@ -0,0 +1,8 @@
+#define DataFormats_TrackerCommon_TrackerDetSide_H
+
+struct TrackerDetSide {
+  enum Side { Barrel = 0, NegEndcap = 1,  PosEndcap = 2 };
```
Is anything else expected inside ```struct TrackerDetSide```?
this could be ```enum class ```

makortel replied:
I thought I had a reason for using unscoped `enum`, but I see in the end there is only one place converting it to/from a number and that is already using `static_cast`. Will change to `enum class` (after you complete the review).

makortel replied:
There were more conversions from the enum to an unsigned int in `HitExtractor*`. Something that could be also considered is to propagate this enum to `TrackerTopology` (as the numbering is intentionally consistent with `TrackerTopology::side()` (but that is beyond this PR).

slava77 commented on RecoTracker/TkSeedingLayers/src/HitExtractorSTRP.h:
```diff
@@ -44,7 +44,7 @@ class HitExtractorSTRP final : public HitExtractor {
   bool skipThis(DetId id, OmniClusterRef const& clus, edm::Handle<edm::ContainerMask<edmNew::DetSetVector<SiStripCluster> > > & stripClusterMask) const;
 
-  void setNoProjection() const {failProjection=true;};
+  void setNoProjection() {failProjection=true;};
```
since this is touched, the trailing ";" can be removed

slava77 commented on RecoTracker/TkSeedingLayers/src/SeedingLayerSetsBuilder.cc:
```diff
@@ -327,50 +347,22 @@ void SeedingLayerSetsBuilder::updateEventSetup(const edm::EventSetup& es) {
 
-SeedingLayerSets SeedingLayerSetsBuilder::layers(const edm::EventSetup& es)
-{
+#include "RecoTracker/TransientTrackingRecHit/interface/TkTransientTrackingRecHitBuilder.h"
```
there is an include of this file already at L12

slava77 commented:
@makortel 
at a quick glance it looks like the updated implementation may take a different time to execute (e.g. SeedingLayerSetsBuilder::hits or SeedingLayersEDProducer::produce).
Do you have some profiler checks to clarify?

slava77 commented:
> Will change to enum class (after you complete the review).

I'm done with the visual inspection

makortel commented:
> at a quick glance it looks like the updated implementation may take a different time to execute (e.g. SeedingLayerSetsBuilder::hits or SeedingLayersEDProducer::produce).
> Do you have some profiler checks to clarify?

I did not profile as I'd expect about the same execution time. The "heavy" computations should be exactly the same with the new code as with the old one, just structured a bit differently. E.g. the old memory allocations (`indices.reserve(theLayers.size())`, `std::move(tmp.begin(), tmp.end(), std::back_inserter(hits))`)  are now in `SeedingLayerSetsHits` constructor and in `addHits()`. If you have further concerns, could you be more specific?

I see that the old `hits.shrink_to_fit()` in `SeedingLayersEDProducer` has no equivalent call, I'll add it back (inside SeedingLayerSetsBuilder::hits()).

makortel committed:
Change TrackerDetSide to enum class

Maybe we need to consider propagating TrackerDetSide to TrackerTopology interface?

makortel committed:
Remove spurious ;

makortel committed:
Remove duplicated #include

makortel committed:
Add back shink_to_fit

makortel commented:
Review comments addressed.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21138/1890

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 3e37411b35a8b426203e4207b3ff09f0e6752379
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21138/24302/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/0493302eeea74843ce15b72c0b162c5ce883537c
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21138/24302/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21138/24302/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21138/24302/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 27
 - DQMHistoTests: Total histograms compared: 2967244
 - DQMHistoTests: Total failures: 32
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2967034
 - DQMHistoTests: Total skipped: 178
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram bins added: 0.0 ( 23 files compared)
 - Checked 111 log files, 8 edm output root files, 27 DQM output files

slava77 commented:
convStep and conversionStep validation is apparently getting lost in the layer set indexing

e.g. in wf 10024
it looks like what was in TIB[34] was moved to "unknown"
[red is with this PR]
<img width="828" alt="wf10024_tk_vs_seedlayerset" src="https://user-images.githubusercontent.com/4676718/32649973-5e0ed4ae-c5fc-11e7-9deb-0bce1a8c5340.png">

The same histogram that was in red above plotted by itself
<img width="759" alt="wf10024_tk_vs_seedlayerset_new" src="https://user-images.githubusercontent.com/4676718/32649999-71bbb0b2-c5fc-11e7-861d-a32851f50833.png">

It looks like some fix is needed.

makortel commented:
> convStep and conversionStep validation is apparently getting lost in the layer set indexing

Thanks @slava77. This indeed the case and happens because I forgot how hacky my solution for "MTXX" in MTV was
https://github.com/cms-sw/cmssw/blob/c92d92a565aa09f960f0bce4963a6af1f5c94449/Validation/RecoTrack/src/MTVHistoProducerAlgoForTracker.cc#L1592-L1600
so basically it is recognized only for strip triplet seeds and for them only for the last hit (as that was the case at the time of implementation within PixelLessStep and TobTecStep).

I agree with this PR the method would need some generalization. It could be sufficient to just remove the "strip triplet+last hit" checks and just map `subdetStrip && trackerHitRTTI::isSingle()` => "MTXX" as that's basically what the added `SeedingLayersEDProducer::fillDescriptions()` suggests (won't work for the `CombinatorialSeedGeneratorForBeamHalo_cfi` where I re-used "MTEC" for other purpose instead of adding a new "layer type", but on the other hand we don't run MTV there). I'll give it a quick try.

makortel commented:
>  It could be sufficient to just remove the "strip triplet+last hit" checks

This seems to work otherwise except for TEC (in conversions) because 
https://github.com/cms-sw/cmssw/blob/c92d92a565aa09f960f0bce4963a6af1f5c94449/RecoTracker/ConversionSeedGenerators/python/ConversionStep_cff.py#L182-L192
because it is configured to in addition to matched hits take the unmatched rphi+stereo hits.

I'm not sure what do here. Clearly this use breaks the assumptions in the validation code that one can obtain the seeding layer (as in the `SeedingLayersEDProducer` configuration) only from the hit itself. On the other hand duplicating the seeding layers for conversion (for "TEC" and "MTEC" variants) does not appeal to me either. A possible solution would be to add an index of the seeding layer set in the configuration to the `TrajectorySeed` (*) and use that in the MTV. Actually for conversions this would be more interesting anyway because the seeds have only the hit from the first layer of the layer pair so there is no way to find out the original pair of layers in the configuration (see #20176).

(*) An index was one possibility on the table at the time of #14248, but we ended up with a simpler  solution (though less general as we see now)

Anyway, I propose I include the quick fix I did here, and leave the further development a subsequent PR as it will take some time. Thoughts, comments?

slava77 commented:
On 11/10/17 11:57 AM, Matti Kortelainen wrote:
> Anyway, I propose I include the quick fix I did here, and leave the
> further development a subsequent PR as it will take some time. Thoughts,
> comments?

My guess is that it is enough to categorize it to be in the right
detector (not necessarily in the exact layer), instead of "unknown".

makortel commented:
Actually I can patch the layer list given to MTV to include `MTEC<n>` for each `TEC<n>`.

makortel committed:
Streamline the strip mono hit definition in MTV

With the recent changes to SeedingLayerSetsBuilder we have MTXX
seeding layers also elsewhere than just the last layer of strip
triplets.

makortel committed:
Hack for conversion seeding layers

makortel committed:
Fix typo

cmsbuild commented:
The code-checks are being triggered in jenkins.

makortel commented:
With this update we minimize the regression in the conversion seeding layer plots, that is TEC bins get split to TEC+MTEC (in addition to some TIB/TOB moving to MTIB/MTOB)
![image](https://user-images.githubusercontent.com/33993/32660055-a12136c2-c621-11e7-8f9f-36088159abc8.png)

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21138/1924

cmsbuild commented:
Pull request #21138 was updated. @perrotta, @vazzolini, @kmaeshima, @dmitrijus, @cmsbuild, @jfernan2, @slava77, @vanbesien can you please check and sign again.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 733226a9234a1a3a8e24c0f397f26c250610b5cc
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21138/24345/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/3eecb4fef06b51670ce2b2a8130985bf87be8ca6
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21138/24345/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21138/24345/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21138/24345/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 27
 - DQMHistoTests: Total histograms compared: 2832699
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 101
 - DQMHistoTests: Total successes: 2832419
 - DQMHistoTests: Total skipped: 178
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 10.04 KiB( 23 files compared)
 - Checked 111 log files, 8 edm output root files, 27 DQM output files

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/21138	733226a
- changes are in line with the PR description. No changes in reco quantities. Some seedingLayer index reshuffling (renaming and split of TEC to TEC+MTEC) in MTV conversion plots.
- jenkins tests pass and comparisons with baseline show only expected differences [ignoring 8 reco diffs in logErrorHarvester in PU workflows apparently due to file re-open]

Here is an example from 10224 concersionStep 
top is baseline, bottom is with this PR
<img width="1094" alt="wf10224_conversionstep_seedlay_tb" src="https://user-images.githubusercontent.com/4676718/32688916-1a99c11e-c6da-11e7-9ace-2a739b5f05c4.png">
the "TEC+MTEC" adds up to what was there in TEC before.

dmitrijus commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar (and backports should be raised in the release meeting by the corresponding L2)

davidlange6 commented:
+1
